src/enum.lisp

```diff
@@ -37,59 +47,138 @@
+
+(defun parse-foreign-enum-like (type-name base-type values
+                                &optional field-mode?)
+  (let ((keyword-values (make-hash-table :test 'eq))
+        (value-keywords (make-hash-table))
+        (field-keywords (list))
+        (bit-index->keyword (make-array 0 :adjustable t
+                                        :element-type t))
+        (default-value (if field-mode? 1 0))
+        (largest-value 0))
     (dolist (pair values)
-      (destructuring-bind (keyword &optional (value default-value))
+      (destructuring-bind (keyword &optional (value default-value value-p))
           (ensure-list pair)
-        (check-type keyword keyword)
+        (check-type keyword enum-key)
         (check-type value integer)
-        (if (gethash keyword (keyword-values type))
+        (when (> value largest-value)
+          (setf largest-value value))
+        (if field-mode?
+            (if value-p
+                (when (and (>= value default-value)
+                           (single-bit-p value))
+                  (setf default-value (ash value 1)))
+                (setf default-value (ash default-value 1)))
+            (setf default-value (1+ value)))
+        (if (gethash keyword keyword-values)
             (error "A foreign enum cannot contain duplicate keywords: ~S."
                    keyword)
-            (setf (gethash keyword (keyword-values type)) value))
-        ;; This is completely arbitrary behaviour: we keep the last we
+            (setf (gethash keyword keyword-values) value))
+        ;; This is completely arbitrary behaviour: we keep the last
         ;; value->keyword mapping. I suppose the opposite would be
         ;; just as good (keeping the first). Returning a list with all
         ;; the keywords might be a solution too? Suggestions
         ;; welcome. --luis
-        (setf (gethash value (value-keywords type)) keyword)
-        (setq default-value (1+ value))))
-    type))
+        (setf (gethash value value-keywords) keyword)
+        (when (and field-mode?
+                   (single-bit-p value))
+          (let ((bit-index (1- (integer-length value))))
+            (push keyword field-keywords)
+            (when (<= (array-dimension bit-index->keyword 0)
+                      bit-index)
+              (setf bit-index->keyword
+                    (adjust-array bit-index->keyword (1+ bit-index)
+                                  :initial-element nil)))
+            (setf (aref bit-index->keyword bit-index)
+                  keyword)))))
+    (if base-type
+        (progn
+          (setf base-type (canonicalize-foreign-type base-type))
+          (assert (member base-type +valid-enum-base-types+ :test 'eq) ()
+                  "Invalid base type ~S for enum type ~S. Must be one of ~S."
+                  base-type type-name +valid-enum-base-types+))
+        ;; details: https://stackoverflow.com/questions/1122096/what-is-the-underlying-type-of-a-c-enum
```

Comment:
nope.

i just bit the bullet and learned about the relevant parts of the C standard, because i was too confused at the beginning. and if i got to know this anyway, then i thought i'll just implement it in CFFI, too.